XI2 mouse jumping fixes

jascrain opened the pull request:
### Purpose of this PR

When running under Wayland on Linux, the mouse position resets whenever the scroll wheel or a mouse button is used. See anegostudios/VintageStory-Issues#381. This fixes two of the reasons this can happen.

### Testing status

These patches have been used by the VintageStory game to improve mouse support in Wayland + Linux.

### Comments

This is not a 100% fix for Wayland + Linux mouse issues. There are still issues that I believe are related to multi-threading in XI2MouseKeyboard, and another that occurs when many mouse events are received at once.

NogginBops commented on src/OpenTK/Platform/X11/XI2MouseKeyboard.cs:
```diff
@@ -397,6 +399,14 @@ private void UpdateDevices()
                     XI.FreeDeviceInfo((IntPtr)list);
                 }
+
+                // Resture current mouse position
```
Might be a slight typo, I think it should be `Restore` and not `Resture`.

jascrain replied:
yes, a typo

NogginBops commented on src/OpenTK/Platform/X11/XI2MouseKeyboard.cs:
```diff
@@ -397,6 +399,14 @@ private void UpdateDevices()
                 }
+
+                // Resture current mouse position
+                if (devices.Count >= 1) {
+                    devices[0].State.Position = current_mouse_state.Position;
+                    devices[0].State.SetScrollAbsolute(
+                            current_mouse_state.Scroll.X,
+                            current_mouse_state.Scroll.Y);
+                }
```
Why do we need to reset the position here?
Is the 0th device special or what is it that is going on?
Are we applying events from other devices to the 0th device?

jascrain replied:
> Why do we need to reset the position here?

This function, UpdateDevices(), is called whenever a DeviceChanged event is received. This function removes and re-adds all devices and thus resets the current mouse position to the default (0, 0). The intent of this is to save and restore the current mouse position instead of resetting to (0, 0).

> Is the 0th device special or what is it that is going on?

The 0th device is not really special here except that I needed to save the position somewhere and it was the most convenient place. My thinking regarding this:

The IMouseDriver2.GetState() function merges all of the state information together, so putting them into a single device is more-or-less equivalent.

I'm not able to find a system that actually supports multiple mouse devices. As far as I can tell, both Xorg and XWayland emulate a single mouse even when multiple mice are present so I don't think there's any regression caused by doing it this way.

jascrain replied:
Instead of this "Save mouse state" patch, I do have an alternate approach I'm working on that you may find cleaner. Instead of calling UpdateDevices() on a DeviceChanged event and saving/restoring the position, it processes the DeviceChanged event and updates the configuration of that device.

NogginBops replied:
From what I understand from your explanation I think the keeping of the position and scroll info is a quite efficient way to solve the problem.
I like that the risk of this change affecting other parts of the software are quite low.

From the sound of your cleaner solution it might require more complicated logic that is harder to prove.
But I might be wrong in that assessment, and I'm also quite fond of "correct" solutions so I think it could be worth a look if it's not a ton of work on your part.

I just want to say I appreciate you taking the time to make this PR. It's helps a lot.

NogginBops commented on src/OpenTK/Platform/X11/XI2MouseKeyboard.cs:
```diff
@@ -541,36 +551,41 @@ private unsafe static void ProcessRawMotion(XIMouse d, ref XIRawEvent raw)
             if (d.MotionX.number != -1)
             {
-                x = ReadRawValue(ref raw, d.MotionX.number);
+                if (ReadRawValue(ref raw, d.MotionX.number, out x))
+                {
+                    if (d.MotionX.mode == XIMode.Relative)
+                    {
+                        d.State.X += (int)Math.Round(x);
+                    }
+                    else
+                    {
+                        d.State.X = (int)Math.Round(x);
+                    }
+                }
```
To me it isn't totally clear from the code why this is a fix for the issue (I'm not very familiar with XI2).
Do you have an explanation for why this fixes the issue?

jascrain replied:
XI2 mouse messages can contain a lot of state information, including buttons, scroll wheels, position, etc., but not all of them are present in each message. Which ones are present is specified in the message's valuators mask, a bitfield.

Previously, ReadRawValue() was just returning 0 when a valuator was not present, which was another way the mouse position was resetting to 0. This "Skip missing valuators" patch changes the ReadRawValue() function to return true or false depending on whether the valuator is present. Only if the valuator is present does this then set the position or scrollwheel.

NogginBops replied:
Yeah ok, I was getting something like that from the code.

So basically, what was happening is that we read the `x` and `y` values without caring if they where there or not and defaulting them to `0` if they where not there.
Then later we just checked if the event was relative or absolute and just applied that `0` blindly.
And the issue then would have been messages that didn't contain a valid `x` or `y` value but where marked as `absolute` and would then set the position to `0`?
(relative would not have been a problem as adding `0` wouldn't change the value)

Am I correct in that understanding?

jascrain replied:
Yes, that's correct.

And this happens more frequently when running on Wayland because it switches to relative mode when moving the mouse, and to absolute mode when clicking or scrolling.

NogginBops reviewed (approved):
These are some great changes!
We definitely want this to get into a opentk3 release.

I have a few questions that would be great to have more insight about.

tyronx commented:
I can confirm that these fixes are working and in productive use.

NogginBops commented:
I don't doubt that these changes work, I'm convinced they do work. I'm happy to merge the changes. 
But it would be nice to get some clarification on why these changes fixes the issue.

tyronx commented:
> I don't doubt that these changes work, I'm convinced they do work. I'm happy to merge the changes.
> But it would be nice to get some clarification on why these changes fixes the issue.

Certainly, just wanted to give my stamp of approval. Kindly request to merge this if no further improvements are made on the PR though.

NogginBops commented:
From my understanding @jascrain intends to change the implementation of `UpdateDevices()`?
But I could be wrong about that. I'm happy to merge this as it is though.

tyronx commented:
> 
> 
> From my understanding @jascrain intends to change the implementation of `UpdateDevices()`?
> But I could be wrong about that. I'm happy to merge this as it is though.

Yea, let's see.

leezer3 commented:
As another 3.x user, I definitely appreciate these, thanks.

Doesn't affect me quite so badly as we don't really use the mouse by default, but I've still got no intentions of migrating to 4x.

jascrain commented:
I've added a "Process DeviceChangedEvent" commit that instead of resetting and querying all of the devices when a DeviceChangedEvent is received, just updates the device that has changed.

I've changed this to a draft because with this, the "Save mouse state across DeviceChanged" commit will not be needed.

NogginBops commented:
@jascrain This is a ping for this issue. It would be really nice to get this merged.
If you don't have that much time to work on this atm then I think we can at least merge the commits before the "Process DeviceChangedEvent" change?

leezer3 commented:
If I read it right, the *last* two commits can be cherry-picked into the tree.

The final commit alters the approach, by correctly processing the DeviceChanged event so that it no longer triggers the bug the first works around.
TBQH, you could probably leave all three in there for a belt and braces approach. (in case any other events trigger said behaviour)

NogginBops commented:
Removing this from the `3.3.3` milestone as it doesn't seem like this PR is going to be finished for that version.

NogginBops commented:
@jascrain guessing you don't mean to finish this PR?
